akka-actor/src/main/java/akka/dispatch/AbstractBoundedNodeQueue.java

```diff
@@ -167,9 +167,17 @@ public final T poll() {
             final Node<T> deq = getDeq();
             final Node<T> next = deq.next();
             if (next != null) {
+                //Can't rely on next.value after casDeq because 'next' can be popped by another thread already
+                //before this one finishes. 
+                //Thus reading and storing the value before casDeq is needed. 
+                T nextVal = next.value;
                 if (casDeq(deq, next)) {
-                    deq.value = next.value;
-                    next.value = null;
+                    deq.value = nextVal;
+                    
+                    //Only set this to null if next still contains the expected
+                    //value. If not, "deq.value = nextVal" has already happened on a different thread.
+                    //and we don't want to override that since that's another value popping.
+                    next.casNullValue(nextVal);
```

Comment:
This does not work since the queue permits the same object to be enqueued multiple times, which means that the CAS would succeed when it shouldn’t.